contracts/MultiSender.sol

```diff
@@ -6,69 +6,122 @@ import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
 import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
 import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
 import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
 
 // Use SafeERC20 library for IERC20Upgradeable
 using SafeERC20 for IERC20Upgradeable;
 
 // Define the contract
-contract MultiSender is Initializable {
+contract MultiSender is Initializable, ReentrancyGuardUpgradeable {
     // Initialize the contract
-    function initialize() public initializer {}
+    function initialize() public initializer {
+        __ReentrancyGuard_init();
+    }
 
     // Event definitions
     event SendETH(address[] recipients, uint256[] amounts);
     event SendERC20(address token, address[] recipients, uint256[] amounts);
 
     // Function to send ETH to multiple recipients
-    function sendETH(address[] calldata _recipients, uint256[] calldata _amounts) external payable {
+    function sendETH(
+        address[] calldata _recipients,
+        uint256[] calldata _amounts
+    ) external payable nonReentrant {
         // Check if the lengths of recipients and amounts arrays are equal
         require(_recipients.length == _amounts.length, "Must have the same length");
 
         // Calculate the total amount to be sent
-        uint256 totalAmount = 0;
+        uint256 _totalAmount = 0;
         for (uint256 i = 0; i < _amounts.length; i++) {
-            totalAmount += _amounts[i];
+            _totalAmount += _amounts[i];
         }
 
+        // Check if the total amount is greater than zero
+        require(_totalAmount > 0, "Total transfer amount is zero");
+
         // Check if the sender has enough ETH
-        require(msg.value == totalAmount, "Unequal transfer amount");
+        require(msg.value == _totalAmount, "Unequal transfer amount");
+
+        // Calculate the failed transaction amount
+        uint256 _failedAmount = 0;
+
+        // Store the initial balance of the contract
+        uint256 initialBalance = address(this).balance - msg.value;
 
         // Send ETH to recipients
         for (uint256 i = 0; i < _recipients.length; i++) {
-            payable(_recipients[i]).transfer(_amounts[i]);
+            // Check if the recipient is the zero address
+            if (_recipients[i] == address(0)) {
+                // Accumulate failed transaction amount
+                _failedAmount += _amounts[i];
+                continue;
+            }
+
+            // Skip the transaction if the amount is zero
+            if (_amounts[i] == 0) continue;
+
+            (bool success, ) = payable(_recipients[i]).call{value: _amounts[i]}("");
+            if (!success) {
+                // Accumulate failed transaction amount
+                _failedAmount += _amounts[i];
+            }
```

Comment:
@Zena-park Do you think it is a good idea to add a check if the ` if (_recipients[i]`  != `this.address`) to avoid sending Eth to the contract itself? Likely I can also add this check to `SendERC20` function.

Reply:
Checking every address for null addresses is very burdensome for gas costs.

Please consult with the project team members, re-check the contract specifications, and if the specifications have not yet been determined, discuss them and make a decision.
Please write the determined specifications in the document (readme). (When developing on the front end or calling the contract directly, the situations to be considered must be specified in detail.)

When determining the specifications, please also specify that duplicate addresses are not considered in the contract, so if you enter the same address multiple times, it will be transmitted inefficiently multiple times. These should be mentioned so that they can be considered during front-end development.

Reply:
@Zena-park We agreed to move a very basic checks to frontend side before call the method. I will remove the `address(0)` check to safe gas and will add the more detailed information to the `README`.

Reply:
@Zena-park We agreed to move a very basic checks to frontend side before call the method. I will remove the address(0) check to safe gas and will add the more detailed information to the README.